scripts/west_commands/runners/mdb.py

```diff
@@ -0,0 +1,104 @@
+
+
+from os import path
+from runners.core import ZephyrBinaryRunner, RunnerCaps
+
+class MdbBinaryRunner(ZephyrBinaryRunner):
+    '''Runner front-end for mdb.'''
+
+    def __init__(self, cfg, cores=1, jtag='digilent', nsim_args='',
+                dig_device=''):
+        super(MdbBinaryRunner, self).__init__(cfg)
+        self.jtag = jtag
+        self.cores = int(cores)
+        if nsim_args != '':
+            self.nsim_args = path.join(cfg.board_dir, 'support', nsim_args)
+        else:
+            self.nsim_args = ''
+        self.elf_name = cfg.elf_file
+        if dig_device != '':
+            self.dig_device = '-prop=dig_device=' + dig_device
+        else:
+            self.dig_device = ''
+        self.commander = 'mdb'
+
+    @classmethod
+    def name(cls):
+        return 'mdb'
+
+    @classmethod
+    def capabilities(cls):
+        return RunnerCaps(commands={'flash', 'debug'})
+
+    @classmethod
+    def do_add_parser(cls, parser):
+        parser.add_argument('--jtag', default='digilent',
+                            help='choose the jtag interface')
+        parser.add_argument('--cores', default=1,
+                            help='choose the cores to debug')
+        parser.add_argument('--nsim_args', default='',
+                            help='if given, arguments for nsim simulator')
+        parser.add_argument('--dig-device', default='',
+                            help='''choose the digilent device when
+                             --jtag=digilent''')
+
+    @classmethod
+    def create(cls, cfg, args):
+        return MdbBinaryRunner(
+            cfg,
+            cores=args.cores,
+            jtag=args.jtag,
+            nsim_args=args.nsim_args,
+            dig_device=args.dig_device
+            )
+
+    def do_run(self, command, **kwargs):
+        self.require(self.commander)
+
+        mdb_basic_options = ['-nooptions', '-nogoifmain',
+                            '-toggle=include_local_symbols=1']
+
+        # nsim
+        if self.nsim_args != '':
+            mdb_target = ['-nsim', '@' + self.nsim_args]
+        # hardware target
+        else:
+            if self.jtag == 'digilent':
+                mdb_target = ['-digilent', self.dig_device]
+            else:
+                # \todo: add support of other debuggers
+                mdb_target = ['']
+
+        if command == 'flash':
+            mdb_run = ['-run', '-cmd=-nowaitq run', '-cmd=quit','-cl']
+        elif command == 'debug':
+            # use mdb gui to debug
+            mdb_run = ['-OKN']
+
+        if self.cores == 1:
+            # single core's md command is different with multicores
+            mdb_cmd = ([self.commander] + mdb_basic_options + mdb_target +
+                       mdb_run + [self.elf_name])
+        elif self.cores <= 4:
+            mdb_multifiles='-multifiles='
+            for i in range(self.cores):
+                mdb_sub_cmd = ([self.commander] +
+                            ['-pset={}'.format(i+1),
```

Comment:
I would strongly suggest to add a comment here as one day some good developer will "fix" `pset` value calculation (from `i+1` to obviously "correct" `i`) and this way very subtly will break our SMP execution. The [sad] fact is `pset` starts indexing from `1` but not `0` and that's important to take into account. Moreover MDB happily accepts `-pset=0` and still refers to the first core in the CMPD ("Concurrent Multi-processor Debugging" as it is referred in the MDB's manual and configuration GUI) set :)

That said we need to keep indexing of `pset` that starts from 1!

Reply:
good idea, i will add comments